keps/sig-scheduling/4120-pod-topology-spread-persisten-volume-claim/README.md

```diff
@@ -0,0 +1,851 @@
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+When calculating pod topology spread skew, I want to exclude nodes that have node affinity volume of terminated pod to prevent recreate termianted pods from falling into unexpected Pending state.
+
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+- Checking pod whether exist depend on PersistentVolumeClaim may lead to performance problem, we need to verify this by adding performance tests. If performance problem does exists, we'd like to add the process of determining whether a pod exists during PreFilter.
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+This KEP proposes a method that use `volume.kubernetes.io/selected-node` annotation and ownerReferences
+of `PersistentVolumeClaim` to replace deleted pod participate in `PodTopologySpread` plugin filter and score process.
+
+The `volume.kubernetes.io/selected-node` annotation of `PersistentVolumeClaim`
+was added by kube-scheduler when `StorageClasses` of `PersistentVolumeClaim` volumeBindingMode
+is [WaitForFirstConsumer](https://kubernetes.io/docs/concepts/storage/storage-classes/#volume-binding-mode),  
+such `StorageClasses` like [LocalStorage](https://kubernetes.io/docs/concepts/storage/storage-classes/#local),
+[AWSElasticBlockStore](https://kubernetes.io/docs/concepts/storage/storage-classes/#aws-ebs),
+[GCEPersistentDisk](https://kubernetes.io/docs/concepts/storage/storage-classes/#gce-pd),
+[AzureDisk](https://kubernetes.io/docs/concepts/storage/storage-classes/#azure-disk).
+
+The ownerReferences of `PersistentVolumeClaim` should was added by `StatefulSet` controller, this will need controller-manager
```

Comment:
Is the owner reference mainly used to detect if the PVC is part of a StatefulSet? Could we infer that from the Pod ownerrefernece instead since the PVC ownerreference would require the use of a new optional feature.

I guess another question I have is do we want to restrict this to only StatefulSets? Or anything with a matching owner reference?

Reply:
> Is the owner reference mainly used to detect if the PVC is part of a StatefulSet? Could we infer that from the Pod ownerrefernece instead since the PVC ownerreference would require the use of a new optional feature.
> 

can't query `deleted` pod , so can't use pod ownerrefernece.

Reply:
> I guess another question I have is do we want to restrict this to only StatefulSets? Or anything with a matching owner reference?

In most cases, only stateful services will use pvc. I have also considered supporting other workloads, but I don’t have any good ideas.